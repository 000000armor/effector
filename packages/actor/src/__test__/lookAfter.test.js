//@flow

import {createActor, readAllIncoming} from '..'

test('look after', async() => {
 const actorA = createActor([{count: 0}])
 const actorB = createActor([{done: 0, fail: 0}]).after(
  ([state], e) => {
   switch (e) {
    case 'fail':
     return {done: false, state}
    case 'done':
     return {done: true, state}
    default:
     return
   }
  },
  async(ctx, e, that) => {
   if (ctx.done) ctx.state.done += 1
   else ctx.state.fail += 1
  },
 )

 actorB.lookAfter(async(err, e, that) => {
  console.log(e)
  console.log(err)
  await that.sendMessage('fail', that)
 }, actorA)

 actorA.after(
  (_, e) => {
   if (e === 'run fail')
    return () => Promise.reject(new Error('just as planned'))
   if (e === 'run done') return () => Promise.resolve('ok')
  },
  async(fn, e, that) => {
   const isFail = {}
   let realError
   const result = await fn().catch(async err => {
    // console.error(err)
    realError = err
    await actorB.sendMessage('fail')
    return isFail
   })
   if (result === isFail) return
   await actorB.sendMessage('done')
   that.state[0].count += 1
  },
 )
 await readAllIncoming([actorA, actorB])

 expect({
  state: {
   a: actorA.state[0],
   b: actorB.state[0],
  },
  meta: {
   a: actorA.meta,
   b: actorB.meta,
  },
 }).toMatchSnapshot()

 await actorA.sendMessage('run done')
 await readAllIncoming([actorA, actorB])
 await expect(actorA.sendMessage('run fail')).resolves.toBeUndefined()
 await readAllIncoming([actorA, actorB])
 await expect(actorA.sendMessage('run fail')).resolves.toBeUndefined()

 await readAllIncoming([actorA, actorB])

 expect({
  state: {
   a: actorA.state[0],
   b: actorB.state[0],
  },
  meta: {
   a: actorA.meta,
   b: actorB.meta,
  },
 }).toMatchSnapshot()
})
