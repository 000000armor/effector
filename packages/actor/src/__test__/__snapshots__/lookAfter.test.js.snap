// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`look after 1`] = `
Object {
  "meta": Object {
    "a": ActorPrivateMeta {
      "handlers": Set {
        [Function],
      },
      "incoming": Set {},
      "keepBy": Actor {
        "id": ID {
          "id": "2",
        },
        "state": Array [
          Object {
            "done": 0,
            "fail": 0,
          },
        ],
      },
      "keeperFor": WeakMap {},
      "newHandlers": Set {},
      "newIncoming": Set {},
    },
    "b": ActorPrivateMeta {
      "handlers": Set {
        [Function],
      },
      "incoming": Set {},
      "keepBy": null,
      "keeperFor": WeakMap {},
      "newHandlers": Set {},
      "newIncoming": Set {},
    },
  },
  "state": Object {
    "a": Object {
      "count": 0,
    },
    "b": Object {
      "done": 0,
      "fail": 0,
    },
  },
}
`;

exports[`look after 2`] = `
Object {
  "meta": Object {
    "a": ActorPrivateMeta {
      "handlers": Set {
        [Function],
      },
      "incoming": Set {},
      "keepBy": Actor {
        "id": ID {
          "id": "2",
        },
        "state": Array [
          Object {
            "done": 1,
            "fail": 2,
          },
        ],
      },
      "keeperFor": WeakMap {},
      "newHandlers": Set {},
      "newIncoming": Set {},
    },
    "b": ActorPrivateMeta {
      "handlers": Set {
        [Function],
      },
      "incoming": Set {},
      "keepBy": null,
      "keeperFor": WeakMap {},
      "newHandlers": Set {},
      "newIncoming": Set {},
    },
  },
  "state": Object {
    "a": Object {
      "count": 1,
    },
    "b": Object {
      "done": 1,
      "fail": 2,
    },
  },
}
`;
